src/socket/communicate.rs

```diff
@@ -0,0 +1,239 @@
+        ///
+        /// This method blocks until a client tries to connect.
+        pub fn accept<F: 'static>(
+            &self,
+            handler: F,
+        ) -> Result<std::thread::JoinHandle<()>, AcceptError>
+        where
+            F: FnOnce(UnixStream, CommunicationType) -> (),
+            F: std::marker::Send,
+        {
+            // - socket accept
+            let (unix_stream, _) = self.listener.accept().map_err(AcceptError::Accept)?;
+            // - read first message as a `CommunicationType`
+            let comm_type: CommunicationType =
+                ReadWriter::<CommunicationType, ConnectionAccepted>::new(&unix_stream)
+                    .react(self.accept_timeout.clone(), |_| ConnectionAccepted())
+                    .map_err(AcceptError::Message)?;
+            // spawn a thread with the accept handler
+            Ok(std::thread::spawn(move || handler(unix_stream, comm_type)))
+        }
+    }
+
+}
+
+/// Clients that can talk to a `Listener`.
+///
+/// `R` is the type of messages this client reads.
+/// `W` is the type of messages this client writes.
+///
+/// The construction of `Client` is only exported for
+/// the pre-defined interactions with the `Listener` we support.
+pub mod client {
+    use super::*;
+    use std::marker::PhantomData;
+
+    /// A `Client` that can talk to a `Listener`.
+    pub struct Client<R, W> {
+        /// Type of interaction with the `Listener`.
+        comm_type: CommunicationType,
+        /// Connected socket.
+        socket: Option<UnixStream>,
+        /// Timeout for reads/writes.
+        timeout: Timeout,
+        phantom_r: PhantomData<R>,
+        phantom_w: PhantomData<W>,
+    }
+
+    /// Error when talking to the `Listener`.
+    #[derive(Debug)]
+    pub enum Error {
+        /// Not connected to the `Listener` socket.
+        NotConnected,
+        /// Read error or write error.
+        Message(ReadWriteError),
+    }
+
+    /// Error when initializing connection with the `Listener`.
+    #[derive(Debug)]
+    pub enum InitError {
+        /// `connect()` syscall failed.
+        SocketConnect(std::io::Error),
+        /// Handshake failed (write `ConnectionType`, read `ConnectionAccepted`).
+        ServerHandshake(ReadWriteError),
+    }
+
+    // builder pattern for timeouts?
+
+    impl<R, W> Client<R, W> {
+        /// “Bake” a Client, aka set its communication type (and message type arguments).
+        /// Not exported.
+        fn bake(timeout: Timeout, comm_type: CommunicationType) -> Client<R, W> {
+            Client {
+                comm_type,
+                socket: None,
+                timeout,
+                phantom_r: PhantomData,
+                phantom_w: PhantomData,
+            }
+        }
+
+        /// Connect to the `Listener` listening on `socket_path`.
+        pub fn connect(self, socket_path: &SocketPath) -> Result<Client<R, W>, InitError> {
```

Comment:
Hmm maybe `connect` should return a ConnectedClient or something so that `socket` ins't an Option. Not sure. What do you think?

Reply:
Yeah, that would be a more type-safe way to do it. I just couldn’t be bothered in this case, because the module is pretty small and doesn’t expose any way for the user to get this wrong (`bake` is not exposed and `connect` already fills in the socket).